### PR TITLE
ENH: Allow scatter plot to plot objects and datetime type data

### LINKED ISSUE #18755: Scatter requires x column to be numeric
```py
In [30]: df = pd.DataFrame(dict(a=['A', 'B', 'C'], b=[2, 3, 4]))

In [31]: plt.scatter(df['a'], df['b'])
Out[31]: <matplotlib.collections.PathCollection at 0x7f17e17b41d0>
```

This works, and produces this:

figure_1

On the other hand, this doesn't:

```py
In [32]: df.plot.scatter(x='a', y='b')
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
/data/documents/uni/phd/projects/FluxnetTrafficLights/scripts/plots/predictability_plots.py in <module>()                                                                                                     
----> 1 df.plot.scatter(x='a', y='b')

~/miniconda3/envs/science/lib/python3.6/site-packages/pandas/plotting/_core.py in scatter(self, x, y, s, c, **kwds)                                                                                           
   2803         axes : matplotlib.AxesSubplot or np.array of them
   2804         """
-> 2805         return self(kind='scatter', x=x, y=y, c=c, s=s, **kwds)
   2806 
   2807     def hexbin(self, x, y, C=None, reduce_C_function=None, gridsize=None,

~/miniconda3/envs/science/lib/python3.6/site-packages/pandas/plotting/_core.py in __call__(self, x, y, kind, ax, subplots, sharex, sharey, layout, figsize, use_index, title, grid, legend, style, logx, logy, loglog, xticks, yticks, xlim, ylim, rot, fontsize, colormap, table, yerr, xerr, secondary_y, sort_columns, **kwds)                                                                                           
   2625                           fontsize=fontsize, colormap=colormap, table=table,
   2626                           yerr=yerr, xerr=xerr, secondary_y=secondary_y,
-> 2627                           sort_columns=sort_columns, **kwds)
   2628     __call__.__doc__ = plot_frame.__doc__
   2629 

~/miniconda3/envs/science/lib/python3.6/site-packages/pandas/plotting/_core.py in plot_frame(data, x, y, kind, ax, subplots, sharex, sharey, layout, figsize, use_index, title, grid, legend, style, logx, logy, loglog, xticks, yticks, xlim, ylim, rot, fontsize, colormap, table, yerr, xerr, secondary_y, sort_columns, **kwds)                                                                                         
   1867                  yerr=yerr, xerr=xerr,
   1868                  secondary_y=secondary_y, sort_columns=sort_columns,
-> 1869                  **kwds)
   1870 
   1871 

~/miniconda3/envs/science/lib/python3.6/site-packages/pandas/plotting/_core.py in _plot(data, x, y, subplots, ax, kind, **kwds)
   1650         if isinstance(data, DataFrame):
   1651             plot_obj = klass(data, x=x, y=y, subplots=subplots, ax=ax,
-> 1652                              kind=kind, **kwds)
   1653         else:
   1654             raise ValueError("plot kind %r can only be used for data frames"

~/miniconda3/envs/science/lib/python3.6/site-packages/pandas/plotting/_core.py in __init__(self, data, x, y, s, c, **kwargs)
    808             # the handling of this argument later
    809             s = 20
--> 810         super(ScatterPlot, self).__init__(data, x, y, s=s, **kwargs)
    811         if is_integer(c) and not self.data.columns.holds_integer():
    812             c = self.data.columns[c]

~/miniconda3/envs/science/lib/python3.6/site-packages/pandas/plotting/_core.py in __init__(self, data, x, y, **kwargs)
    783             y = self.data.columns[y]
    784         if len(self.data[x]._get_numeric_data()) == 0:
--> 785             raise ValueError(self._kind + ' requires x column to be numeric')
    786         if len(self.data[y]._get_numeric_data()) == 0:
    787             raise ValueError(self._kind + ' requires y column to be numeric')

ValueError: scatter requires x column to be numeric
```

Why does pandas require x to be numeric if matplotlib doesn't?

using versions from conda, on Kubuntu 17.10:

```
matplotlib                2.1.0            py36hba5de38_0  
pandas                    0.20.3                   py36_0  
python                    3.6.3                h0ef2715_3  
```

### LINKED ISSUE #30391: df.plot() with kind='scatter' and datetime on x axis bug
I noticed while helping a student that there's a bug when calling a scatter plot as a method on a DataFrame. Regardless of what the index is or what the x argument is, it appears that pandas makes the x argument into the index, and then searches for the x in the columns, where it no longer exists. When calling the same directly through matplotlib as plt.scatter, it worked properly.

Pandas_Error_scatter

### PR DESCRIPTION
**Pull request by charlesdong1991**
closes #18755 
closes #30391 

xref #8113

- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Because this is a plot PR, so I put the figures gotten by the two examples in those two PRs in this description to facilitate seeing output, and more cases are tested in test files.
Screen Shot 2019-12-23 at 7 34 21 PM

### CODE REVIEW

File: doc/source/whatsnew/v1.0.0.rst [hunk 1]
```diff
@@ -824,6 +824,7 @@ Plotting
 - :func:`set_option` now validates that the plot backend provided to ``'plotting.backend'`` implements the backend when the option is set, rather than when a plot is created (:issue:`28163`)
 - :meth:`DataFrame.plot` now allow a ``backend`` keyword arugment to allow changing between backends in one session (:issue:`28619`).
 - Bug in color validation incorrectly raising for non-color styles (:issue:`29122`).
+- Allow :meth: `DataFrame.plot` to plot ``objects`` and ``datetime`` type data (:issue:`18755`, :issue:`30391`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.0.0.rst]: this is more specific yes? it only applies to scatter right

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v1.0.0.rst (hunk 1 above)
Comment (charlesdong1991) [reply doc/source/whatsnew/v1.0.0.rst]: thanks for taking a look! @jreback indeed, should have been more specific.

Updated.

File: pandas/plotting/_matplotlib/core.py [hunk 2]
```diff
@@ -401,6 +401,10 @@ def _compute_plot_data(self):
             include_type = [np.number]
             exclude_type = ["timedelta"]
 
+        # GH 18755, include object and category type for scatter plot
+        if self._kind == "scatter":
+            include_type.extend(["object", "category"])
```
Comment (WillAyd) [inline pandas/plotting/_matplotlib/core.py]: Just out of curiosity how is the X axis determined for objects? Is there an implicit sort?

Comment (charlesdong1991) [reply pandas/plotting/_matplotlib/core.py]: For `scatter` plot, i think both x and y axis can plot objects data. btw, what do you mean by `implicit sort`?

i think the the order of objects on X axis will be the same with the order of given objects, e.g. 
```python
df = pd.DataFrame({"x": ["b", "z", "c"], "y": [1, 2, 3], })
plt.scatter(x=df["x"], y=df["y"])
```
on X axis, the order will still be b, z, c, other than b, c, z.  And this is identical to behaviour of matplotlib. Does this answer your question? @WillAyd

Comment (WillAyd) [reply pandas/plotting/_matplotlib/core.py]: Could you update or add a test case for mixed object types? Something where the X-axis would be say `["A", 1, "2"]`?

Comment (charlesdong1991) [reply pandas/plotting/_matplotlib/core.py]: emm, but this mixed type is not valid input for scatter plot as far as I know. @WillAyd 

Or do you want to have 
```python
with pytest.raises('something'):
    df.plot(kind='scatter, x='x', y='y')
```

Comment (WillAyd) [reply pandas/plotting/_matplotlib/core.py]: No I think I might just misunderstand the PR. So it's not necessarily just `object` dtypes that can be plotted but `object` dtypes containing only numeric values right?

Comment (charlesdong1991) [reply pandas/plotting/_matplotlib/core.py]: yeah, i just did a quick test, and you could see in the picture attached, a `TypeError` will be raised: `TypeError: 1 is not a string`
Screen Shot 2019-12-24 at 9 49 54 PM

Comment (charlesdong1991) [reply pandas/plotting/_matplotlib/core.py]: oops, did not see your comment when commenting 😅 

Yes, the type has to be same on either axis! maybe the name of PR is misleading, sorry about that. I meant `object` type data (all data remains same type) can be plotted now.

Comment (WillAyd) [reply pandas/plotting/_matplotlib/core.py]: I'm +/- 0 on the `object` dtype here but maybe its OK if there is an implicit assumption that whatever objects provided are sortable. Can you maybe just start by making your test case `["C", "B", "A"]` along the axis or is that not possible?

Comment (charlesdong1991) [reply pandas/plotting/_matplotlib/core.py]: it is definitely possible, i attach the plot here along with output with matplotlib as reference:
Screen Shot 2019-12-24 at 10 02 07 PM

just double check a bit: you want such cases to be added to tests, right?

**Review by WillAyd — APPROVED** [review]
lgtm

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @charlesdong1991

**Pull request merged**